### PR TITLE
Add support for energy dependent spatial models

### PR DESCRIPTION
**Pull request by QRemy**
- TemplateSpatial model support for cubes
- remove TemplateSpatial norm parameter
- Add optional energy argument in spatial model evaluation to support 3d case
- simplify units checks in TemplateSpatial and add _check_unit() method in SkyModel init

### CODE REVIEW

File: gammapy/modeling/models/spatial.py [hunk 1]
```diff
@@ -109,9 +121,12 @@ def position_error(self):
         )
 
     def evaluate_geom(self, geom):
-        """Evaluate model on `~gammapy.maps.Geom`."""
         coords = geom.get_coord(frame=self.frame)
-        return self(coords.lon, coords.lat)
+        energy_name = _get_energy_axis(geom)
+        if energy_name is not None:
+            return self(coords.lon, coords.lat, coords[energy_name])
```
Comment (adonath) [inline gammapy/modeling/models/spatial.py:127]: I think we should rely on numpy broadcasting here, just as we do in `SkyModel.evaluate_geom()`, so evaluate the model not using the full 3D coordinate arrays, but relying on the outer product between the spatial and energy coordinates...

File: gammapy/modeling/models/spatial.py [hunk 2]
```diff
@@ -156,8 +171,11 @@ def plot(self, ax=None, geom=None, **kwargs):
 
         data = self.evaluate_geom(geom)
         m = Map.from_geom(geom, data=data.value, unit=data.unit)
-        _, ax, _ = m.plot(ax=ax, **kwargs)
 
+        if len(m.data.shape) > 2:
```
Comment (adonath) [inline gammapy/modeling/models/spatial.py:175]: Maybe for consistency we just copy the `Map` pattern here? So expose the interactive plotting using a separate `.plot_interactive()` method? I can see that merging results in a more compact API, but I think consistency is more important here...

File: gammapy/modeling/models/cube.py [hunk 3]
```diff
@@ -91,6 +92,26 @@ def _check_covariance(self):
                 [model.covariance for model in self._models],
             )
 
+    def _check_unit(self):
+        from astropy.time import Time
+        from gammapy.data.gti import GTI
+
+        # evaluate over a test geom to check output unit
+        # TODO simpler way to test this ?
```
[the remark this answers is not in the document]
Comment (QRemy) [reply gammapy/modeling/models/cube.py:100]: I tried but the evaluate method is missing for PointSpatialModel

[the remark this answers is not in the document]
Comment (adonath) [reply gammapy/modeling/models/cube.py:100]: Indeed you are right, so just keep as is...

File: gammapy/modeling/models/spatial.py (hunk 1 above)
Comment (QRemy) [reply gammapy/modeling/models/spatial.py:127]: Is this working for a full 3d template model ? SkyDiffuseCube use the SkyModelBase.evaluate_geom that look like this one, no ?

Comment (adonath) [reply gammapy/modeling/models/spatial.py]: Yes, it should work. In `Map.interp_by_coord()`, we call `MapCoord.create()`, which does the broadcasting of the arrays on init:

```python3
from gammapy.maps import MapAxis, Map
import numpy as np

axis = MapAxis.from_energy_bounds("1 TeV", "10 TeV", nbin=10)

m = Map.create(width=10, axes=[axis])

m.interp_by_coord(
{
    "skycoord": m.geom.to_image().get_coord().skycoord,
    "energy": axis.center.reshape((-1, 1, 1))
}
)
```

**Pull request merged**